src/accelerate/test_utils/scripts/test_script.py

```diff
@@ -55,10 +57,13 @@ def process_execution_check():
     accelerator = Accelerator()
     num_processes = accelerator.num_processes
     with accelerator.main_process_first():
-        idx = torch.tensor(accelerator.process_index).to(accelerator.device)
-    idxs = accelerator.gather(idx)
-    if num_processes > 1:
-        assert idxs[0] == 0, "Main process was not first."
+        if accelerator.is_main_process:
+            time.sleep(0.1)  # ensure main process takes longest
+            open("check_main_process_first.txt", "w").close()
+        assert os.path.isfile("check_main_process_first.txt"), "Main process did not create check file yet."
+    accelerator.wait_for_everyone()
+    if accelerator.is_main_process:
+        os.remove("check_main_process_first.txt")
```

Comment:
Thanks! I like the expansion idea. I should point out a couple of issues though:
- `PartialState().main_process_first()` actually works fine in the current release. The issue is in the delegation of the context manager in `Accelerator` to that in `State` and so on.
- If we switch to `Accelerator` the test doesn't do what we want because the main process writes first due to the time delays, causing the test to succeed with the current release. This is why I had a delay in the main process _before_ creating the file and no delay in other processes, to ensure that the full block of code in the context must be executed on main process before being entered in other processes. I hope I'm thinking correctly; does that makes sense? :)

I'll try to modify the test to incorporate this idea tomorrow.

Reply:
Modified the test to incorporate the suggestion. Confirmed that it currently fails on 0.18.0 and passes with the fix. Should be ready for review.